src/Microsoft.AspNetCore.SignalR.Core/Internal/KeepAliveChannel.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Core.Internal
+{
+    internal class KeepAliveChannel : Channel<HubMessage>
```

Comment:
I'm not too sure about this..

Reply:
The thing I like about it is that it keeps the consumption code clean. Ideally, you really want to put keep-alive code in the `HubEndPoint.OnConnectionAsync.WriteToTransport` local function, since that's the central place messages appear. We could have a background task firing pings, but the goal of the ping is to send one if the timeout elapses while waiting for a message. A fixed interval kinda achieves that, but in a weird way and requires synchronization in order to know the time since the last message.

I tried putting this logic into the consumption side (in `WriteToTransport`) and it gets complicated fast, because KeepAlive isn't always enabled. Though I have another idea I'm going to try.

Reply:
This class as a concept or name??

Reply:
Probably the concept, it's a little sketchy and I'm not super happy with it myself.